cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaAssemblyOperator.java

```diff
@@ -1169,6 +1220,41 @@ private boolean isPodToRestart(StatefulSet ss, Pod pod, boolean isAncillaryCmCha
             return !isPodUpToDate || !isPodCaCertUpToDate || isAncillaryCmChange || isCaCertsChanged;
         }
 
+        private boolean isMaintenanceTimeWindowsSatisfied() {
+            try {
+                boolean isSatisfiedBy = maintenanceTimeWindows == null || maintenanceTimeWindows.isEmpty();
+                if (!isSatisfiedBy) {
+                    Date date = new Date();
+                    for (String cron : maintenanceTimeWindows) {
+                        CronExpression cronExpression = new CronExpression(cron);
+                        // the user defines the cron expression in "his" timezone but CO pod
+                        // can be running on a different one, so setting it on the cron expression
+                        cronExpression.setTimeZone(TimeZone.getDefault());
```

Comment:
Maybe I was wrong I should have used `Calendar.getInstance().getTimeZone()` for setting the time zone on the cronexpression.
The idea is that the user write the cron expression in his timezone but of course the TZ on the pod could be different so I force the pod TZ to the cron expression so that it will be evaluate againg the right time and TZ on the pod (while the user still see the cron expression string in his timezone).
The `CronTest.testUserTimeZoneVsPodTimeZone` does something like this.

Reply:
Currently nothing knows what the user's TZ _is_ though; lacking that, what does the cron expression even _mean_? If I ask to you give me a wake up call at 0730 my time, but don't tell you where in the world I am, how can you possibly call me at the right time?

Reply:
>  I pass the "* * 14-15 * * ?" cron expression to the system which applies the pod TZ so for example if it's in the "Pacific/Easter", the cron expression should be evaluated as "* * 8-9 * * ?".

How does that conversion work if you've not told anyone that you're in Italy? The pod doesn't know. The resource doesn't know. What does 'apply' mean? Does it mean that the cron expression is taken to be in that TZ, and it will convert it before comparing with the `Date` when deciding if the expression is satisfied? In which case you should be applying CET/CEST to the expression, not `Pacific/Easter`.

I'm happy to be wrong about this, but right now I just don't understand how it can possibly work without knowing that TZ the user was using when they wrote the expression.

Reply:
I got your point ...
So let's consider the example of my in Italy and I want maintenance every day from 14 to 15.
The pod system is in "Pacific/Easter" zone so that maintenance should main from 8 to 9 (6 hours behind me).
Imagine that I want to test it so I could:

- have a cron expression like  "* * 14-15 * * ?" and then set on the CronExpression object MY TZ so for example "Europe/Rome"
- making a Date like November 26, 8 AM with  "Pacific/Easter" as TZ

so something like this:

```java
public void testUserTimeZoneVsPodTimeZone() throws ParseException {
        // the "user" writes the cron expression in his timezone
        // every second, every minute, hour 14 to 15 every day --> from 14:00 to 15:59
        CronExpression cronExpression = new CronExpression("* * 14-15 * * ?");

        // the pod is running on a "Pacific/Easter" timezone data center, so cron expression needs the right timezone for evaluation
        cronExpression.setTimeZone(TimeZone.getTimeZone(ZoneId.of("Europe/Rome")));

        // it's really 20:00 in "Pacific/Easter" but 14:00 for "user"
        Date d = Date.from(LocalDateTime.of(2018, 11, 26, 8, 00, 0).atZone(ZoneId.of("Pacific/Easter")).toInstant());
        assertTrue(cronExpression.isSatisfiedBy(d));
    }
```

Does it sound reasonable to you?
It really means that we have to provide the user a way to specify the TZ in the maintenance windows property.

Reply:
Saying that the user has to define the time windows in UTC time it's simpler even because we can avoid to add more information to the CRD so I would go for this one. It means that we have to set the "UTC" timezone to the `CronExpression`.

> But either way we need to know the TZ at the point we're evaluating the cron expr.

Why we have to know that? The `new Date()` which is used for validating the cron expression create current Date with the pod timezone but then the cron expression will evaluate that in terms of "UTC" as we set on it.